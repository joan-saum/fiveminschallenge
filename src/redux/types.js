// @flow

import type {Dispatch as ReduxDispatch} from 'redux';

import type {State as S} from './reducer';
export type State = S;

import type {Action as A} from './actions';
export type Action = A;

export type Dispatch = ReduxDispatch<Action>;
