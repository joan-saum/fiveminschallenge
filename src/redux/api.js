// @flow

const API_KEY = '17f5b1b9edaca0b97ee5aa99274079a0';

const buildQueryString = (obj: {}): string => {
  if (!obj || !Object.keys(obj).length) {
    return '';
  }

  return `?${Object.keys(obj)
    .map(k => `${k}=${obj[k]}`)
    .join('&')}`;
};

const api = async function api(route: string, params?: any = {}): Promise<any> {
  const queries = {
    api_key: API_KEY,
    ...params,
  };

  const requestUrl = `https://api.themoviedb.org/3${route}${buildQueryString(
    queries,
  )}`;

  const res = await fetch(requestUrl, {});

  return res.json();
};

export default api;
