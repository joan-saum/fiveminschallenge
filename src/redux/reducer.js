// @flow

import {handle} from 'redux-pack';

import type {ActionReduxPack} from './actions';

export const typedHandle = <T>(
  state: T,
  action: ActionReduxPack,
  handlers: {|
    start?: T => T,
    success?: T => T,
    failure?: T => T,
    finish?: T => T,
    always?: T => T,
  |},
): T => handle(state, action, handlers);

export type MovieType = {|
  image: string,
  title: string,
  overview: string,
  id: number,
|};

export type State = {|
  +items: $ReadOnlyArray<MovieType>,
|};

const IMG_URL = 'https://image.tmdb.org/t/p/original';

const defaultState = {
  items: [],
};

const reducer = (
  state: State = defaultState,
  action: ActionReduxPack,
): State => {
  const {type, payload} = action;

  switch (type) {
    case 'GET_POPULAR_MOVIES':
      return typedHandle(state, action, {
        success: prevState => {
          const {results} = payload;

          const items = results.map(r => ({
            image: `${IMG_URL}${r.poster_path}`,
            title: r.title,
            overview: r.overview,
            id: r.id,
          }));

          return {
            ...prevState,
            items,
          };
        },
      });

    case 'SEARCH_MOVIES':
      return typedHandle(state, action, {
        success: prevState => {
          const {results} = payload;

          const items = results.map(r => ({
            image: `${IMG_URL}${r.poster_path}`,
            title: r.title,
            overview: r.overview,
            id: r.id,
          }));

          return {
            ...prevState,
            items,
          };
        },
      });

    default:
      return state;
  }
};

export default reducer;
