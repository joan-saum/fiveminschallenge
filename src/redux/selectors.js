// @flow

import type {State, MovieType} from './reducer';

export const getMovieById = (state: State, id: number): ?MovieType =>
  state.items.find(item => item.id === id);
