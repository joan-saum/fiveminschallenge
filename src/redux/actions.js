// @flow

import api from './api';

export type Action = {|
  type: string,
  promise: Promise<any>,
  payload: any,
|};

const actions = {
  getPopularMovies: (): Action => ({
    type: 'GET_POPULAR_MOVIES',
    promise: api('/movie/popular'),
    payload: '',
  }),

  searchMovies: (search: string): Action => ({
    type: 'SEARCH_MOVIES',
    promise: api('/search/movie', {query: search}),
    payload: '',
  }),
};

export default actions;
