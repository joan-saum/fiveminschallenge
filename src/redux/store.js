// @flow

import {createStore, applyMiddleware, compose} from 'redux';
import {middleware as reduxPackMiddleware} from 'redux-pack';

import type {State, Action, Dispatch} from './types';
import type {Store} from 'redux';
import reducer from './reducer';

export const store: Store<State, Action, Dispatch> = createStore<
  State,
  Action,
  Dispatch,
>(reducer, compose(applyMiddleware(reduxPackMiddleware)));
