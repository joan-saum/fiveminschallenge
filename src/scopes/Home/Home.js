// @flow

import React, {useEffect, useCallback} from 'react';
import {FlatList} from 'react-native';
import type {Node} from 'react';
import {useSelector, useDispatch} from 'react-redux';

import actions from '../../redux/actions';
import type {MovieType} from '../../redux/reducer';

import type {NavigationProps} from '../../types/navigation';

import MovieItem from '../../components/MovieItem/MovieItem';
import InputSearch from '../../components/InputSearch/InputSearch';

type Props = {|
  navigation: NavigationProps,
|};

const Home = ({navigation}: Props): Node => {
  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(actions.getPopularMovies());
  }, [dispatch]);

  const navigateToMovie = useCallback(
    (id: number) => {
      navigation.navigate('Movie', {id});
    },
    [navigation],
  );

  const renderItem = useCallback(
    ({item}: {item: MovieType}) => {
      return <MovieItem {...item} onPress={navigateToMovie} />;
    },
    [navigateToMovie],
  );

  const keyExtractor = useCallback((item: MovieType) => {
    return item.id.toString();
  }, []);

  const searchMovies = useCallback(
    (search: string) => {
      if (search) {
        dispatch(actions.searchMovies(search));
      } else {
        dispatch(actions.getPopularMovies());
      }
    },
    [dispatch],
  );

  const header = useCallback(() => {
    return <InputSearch onSearch={searchMovies} />;
  }, [searchMovies]);

  const items = useSelector(state => state.items);

  return (
    <FlatList
      data={items}
      renderItem={renderItem}
      ListHeaderComponent={header}
      keyExtractor={keyExtractor}
    />
  );
};

export default Home;
