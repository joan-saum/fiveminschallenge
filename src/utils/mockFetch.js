// @flow

import fetch from 'jest-fetch-mock';
import type {Store, Reducer, StoreEnhancer} from 'redux';
import {createStore, applyMiddleware, compose} from 'redux';
import {middleware as reduxPackMiddleware} from 'redux-pack';

import type {Action, Dispatch} from '../redux/types';

export const mockApiSuccess = (response: any) => {
  fetch.mockResponseOnce(JSON.stringify(response), {
    headers: {
      'Content-Type': 'application/json',
    },
  });
};

export const mockApiFailure = (response: any) => {
  fetch.mockResponseOnce(JSON.stringify(response), {
    status: 555,
    headers: {
      'Content-Type': 'application/json',
    },
  });
};

export const mockStore = <S>(
  reducer: Reducer<S, Action>,
  preloadedState?: S,
): Store<S, Action, Dispatch> => {
  const middlewares: StoreEnhancer<S, Action, Dispatch> = compose(
    applyMiddleware(reduxPackMiddleware),
  );

  return createStore(reducer, preloadedState, middlewares);
};

export const configureTests = () => {
  global.fetch = fetch;
};
