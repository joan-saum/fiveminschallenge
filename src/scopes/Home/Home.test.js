// @flow

import * as React from 'react';
import {Provider} from 'react-redux';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import {FlatList} from 'react-native';

import actions from '../../redux/actions';
import {defaultState} from '../../redux/reducer';

import {configureTests} from '../../utils/mockFetch';

configureTests();

import Home from './Home';

const mockStore = configureStore([]);

jest.unmock('./Home');

const defaultProps = {};

jest.useFakeTimers();

describe('Home', () => {
  it('should render a list of data', () => {
    const item = {id: 1, overview: 'overview', title: 'title', image: 'image'};
    const store = mockStore({
      items: [item],
    });
    const tree = renderer.create(
      <Provider store={store}>
        <Home {...defaultProps} />
      </Provider>,
    );

    expect(tree).toMatchSnapshot();

    const key = tree.root.findByType(FlatList).props.keyExtractor(item);

    expect(key).toBe('1');

    // For the action dispatched
    jest.advanceTimersByTime(1);
    expect(store.getActions()).toStrictEqual([actions.getPopularMovies()]);
  });

  it('should render the header of the list', () => {
    const store = mockStore(defaultState);
    const tree = renderer.create(
      <Provider store={store}>
        <Home {...defaultProps} />
      </Provider>,
    );

    const header = tree.root.findByType(FlatList).props.ListHeaderComponent();

    const headerCreated = renderer.create(header);

    expect(headerCreated).toMatchSnapshot();
  });

  it('should dispatch the action search movies when searching a movie', () => {
    const store = mockStore(defaultState);
    const tree = renderer.create(
      <Provider store={store}>
        <Home {...defaultProps} />
      </Provider>,
    );

    jest.advanceTimersByTime(1);

    const header = tree.root.findByType(FlatList).props.ListHeaderComponent();

    header.props.onSearch('test');
    jest.advanceTimersByTime(1);

    expect(store.getActions()).toStrictEqual([
      actions.getPopularMovies(),
      actions.searchMovies('test'),
    ]);
  });

  it('should dispatch the action get popular movies when entering an empty search', () => {
    const store = mockStore(defaultState);
    const tree = renderer.create(
      <Provider store={store}>
        <Home {...defaultProps} />
      </Provider>,
    );

    jest.advanceTimersByTime(1);

    const header = tree.root.findByType(FlatList).props.ListHeaderComponent();

    header.props.onSearch('');
    jest.advanceTimersByTime(1);

    expect(store.getActions()).toStrictEqual([
      actions.getPopularMovies(),
      actions.getPopularMovies(),
    ]);
  });
});
