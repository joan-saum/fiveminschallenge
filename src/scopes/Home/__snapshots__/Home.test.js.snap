// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home should render a list of data 1`] = `
<RCTScrollView
  ListHeaderComponent={[Function]}
  data={
    Array [
      Object {
        "id": 1,
        "image": "image",
        "overview": "overview",
        "title": "title",
      },
    ]
  }
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  updateCellsBatchingPeriod={50}
  viewabilityConfigCallbackPairs={Array []}
  windowSize={21}
>
  <View>
    <View
      onLayout={[Function]}
    >
      <View />
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View />
    </View>
  </View>
</RCTScrollView>
`;

exports[`Home should render the header of the list 1`] = `<View />`;
